massa-executed-ops/src/processed_denunciations.rs

```diff
@@ -0,0 +1,311 @@
+//! Copyright (c) 2023 MASSA LABS <info@massa.net>
+
+use std::collections::{BTreeMap, HashSet};
+use std::ops::Bound::{Excluded, Included, Unbounded};
+
+use nom::{
+    error::{context, ContextError, ParseError},
+    multi::length_count,
+    sequence::tuple,
+    IResult, Parser,
+};
+
+use crate::{ProcessedDenunciationsChanges, ProcessedDenunciationsConfig};
+
+use massa_hash::{Hash, HASH_SIZE_BYTES};
+use massa_models::streaming_step::StreamingStep;
+use massa_models::{
+    denunciation::{DenunciationIndex, DenunciationIndexDeserializer, DenunciationIndexSerializer},
+    slot::{Slot, SlotDeserializer, SlotSerializer},
+};
+use massa_serialization::{
+    Deserializer, SerializeError, Serializer, U64VarIntDeserializer, U64VarIntSerializer,
+};
+
+const PROCESSED_DENUNCIATIONS_HASH_INITIAL_BYTES: &[u8; 32] = &[0; HASH_SIZE_BYTES];
+
+/// A structure to list and prune previously processed denunciations
+#[derive(Debug, Clone)]
+pub struct ProcessedDenunciations {
+    ///
+    config: ProcessedDenunciationsConfig,
+    /// for better pruning complexity
+    pub sorted_denunciations: BTreeMap<Slot, HashSet<DenunciationIndex>>,
+    /// for better insertion complexity
+    pub denunciations: HashSet<DenunciationIndex>,
+    /// Accumulated hash of the processed denunciations
+    pub hash: Hash,
+    // /// processed status of denunciations (true: success, false: fail)
+    // pub de_processed_status: HashMap<DenunciationIndex, bool>,
+}
+
+impl ProcessedDenunciations {
+    /// Create a new `ProcessedDenunciations`
+    pub fn new(config: ProcessedDenunciationsConfig) -> Self {
+        Self {
+            config,
+            sorted_denunciations: Default::default(),
+            denunciations: Default::default(),
+            hash: Hash::from_bytes(PROCESSED_DENUNCIATIONS_HASH_INITIAL_BYTES),
+            // de_processed_status: Default::default(),
+        }
+    }
+
+    /// Reset the executed operations
+    ///
+    /// USED FOR BOOTSTRAP ONLY
+    pub fn reset(&mut self) {
```

Comment:
we shouldn't be using this for bootstrap: structures should be recreated, see https://github.com/massalabs/massa/issues/3728